addons/account/models/account_report.py

```diff
@@ -0,0 +1,643 @@
+    print_on_new_page = fields.Boolean('Print On New Page', help='When checked this line and everything after it will be printed on a new page.')
+    action_id = fields.Many2one(string="Action", comodel_name='ir.actions.actions')
+    hide_if_zero = fields.Boolean(string="Hide if Zero", help="This line and its children will be hidden when all of their columns are at 0.")
+    domain_formula = fields.Char(string="Domain Formula Shortcut", help="Internal field to shorten expression_ids creation for the domain engine", inverse='_inverse_domain_formula', store=False)
+    account_codes_formula = fields.Char(string="Account Codes Formula Shortcut", help="Internal field to shorten expression_ids creation for the account_codes engine", inverse='_inverse_account_codes_formula', store=False)
+    aggregation_formula = fields.Char(string="Aggregation Formula Shortcut", help="Internal field to shorten expression_ids creation for the aggregation engine", inverse='_inverse_aggregation_formula', store=False)
+
+    _sql_constraints = [
+        ('code_uniq', 'unique (code)', "A report line with the same code already exists."),
+    ]
+
+    @api.depends('parent_id.hierarchy_level')
+    def _compute_hierarchy_level(self):
+        for report_line in self:
+            if report_line.parent_id:
+                report_line.hierarchy_level = report_line.parent_id.hierarchy_level + 2
+            else:
+                report_line.hierarchy_level = 1
+
+    @api.depends('parent_id.report_id')
+    def _compute_report_id(self):
+        for report_line in self:
+            if report_line.parent_id:
+                report_line.report_id = report_line.parent_id.report_id
+
+    @api.constrains('parent_id')
+    def _validate_groupby_no_child(self):
+        for report_line in self:
+            if report_line.parent_id.groupby:
+                raise ValidationError(_("A line cannot have both children and a groupby value (line '%s').", report_line.parent_id.name))
+
+    @api.constrains('expression_ids', 'groupby')
+    def _validate_formula(self):
+        for expression in self.expression_ids:
+            if expression.engine == 'aggregation' and expression.report_line_id.groupby:
+                raise ValidationError(_("Groupby feature isn't supported by aggregation engine."))
+
+    def _copy_hierarchy(self, code_mapping, report=None, copied_report=None, parent=None):
+        ''' Copy the whole hierarchy from this line by copying each line children recursively and adapting the
+        formulas with the new copied codes.
+
+        :param report: The financial report that triggered the duplicate.
+        :param copied_report: The copy of the report.
+        :param parent: The parent line in the hierarchy (a copy of the original parent line).
+        :param code_mapping: A dictionary keeping track of mapping old_code -> new_code
```

Comment:
That parameter is a bit difficult to grok at first glance, and possibly dangerous. Maybe it should be hidden in a "copy state" object so nobody is tempted to pass in external data? From what I understand its role is simply to memo the copy state / previous renamings between line copies, that correct?

Reply:
It's simply the mapping from the old report line code to the new line from my understanding